core/kernel/transfer_list.c

```diff
@@ -389,14 +394,26 @@ bool transfer_list_set_data_size(struct transfer_list_header *tl,
 		    tl->size + mov_dis > tl->max_size) {
 			return false;
 		}
-		ru_new_ev = old_ev + mov_dis;
-		memmove((void *)ru_new_ev, (void *)old_ev, tl_old_ev - old_ev);
+		r_new_ev = old_ev + mov_dis;
 		tl->size += mov_dis;
-		gap = ru_new_ev - new_ev;
 	} else {
-		gap = old_ev - new_ev;
+		/*
+		 * Move distance should be round down to meet the requirement of
+		 * transfer entry data max alignment.
+		 */
+		mov_dis = ROUNDDOWN(old_ev - new_ev,
+				    TL_ALIGNMENT_FROM_ORDER(tl->alignment));
+		r_new_ev = old_ev - mov_dis;
+		tl->size -= mov_dis;
 	}
+	/* Move all following entries to fit in the expanded or shrunk space */
+	memmove((void *)r_new_ev, (void *)old_ev, tl_old_ev - old_ev);
 
+	/*
+	 * Fill the gap due to round up/down with a valid entry if the size of
+	 * the gap is more than an entry header.
```

Comment:
What if the next entry is a `TL_TAG_EMPTY`? Would it make sense to merge the two entries?

Reply:
I guess, idealy we need another API for fragment management.
But I think we can skip it in this PR, since currently when an entry is marked with void' it will be never re-used.
`transfer_list_add()` will just append new entries in the tail (The spec doesn't mandatorily require reuse of 'void' entries, it is still a question whether we need to support the reuse).

Reply:
What the spec says about this isn't so important, but I agree that we can save it for another day.